File: app/src/main/java/edu/pokemon/iut/pokedex/data/model/PokemonTypeStringTypeConverters.java
package edu.pokemon.iut.pokedex.data.model;

import android.arch.persistence.room.TypeConverter;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.util.Collections;
import java.util.List;

public class PokemonTypeStringTypeConverters {
    
    static Gson gson = new Gson();
    
    @TypeConverter
    public static List<Type> stringToSomeObjectList(String data) {
        if (data == null) {
            return Collections.emptyList();
        }

        java.lang.reflect.Type listType = new TypeToken<List<Type>>() {}.getType();

        return gson.fromJson(data, listType);
    }

    @TypeConverter
    public static String someObjectListToString(List<Type> someObjects) {
        return gson.toJson(someObjects);
    }
}

File: app/src/main/java/edu/pokemon/iut/pokedex/data/model/PokemonTypeStringTypeConverters.kt
package edu.pokemon.iut.pokedex.data.model

import android.arch.persistence.room.TypeConverter

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

import java.util.Collections

/**
 * [TypeConverter] are used to convert some complex object in a way that [android.arch.persistence.room.RoomDatabase] can store.<br></br>
 * This class allow to convert Any complex [<] in Json to store in the database, and to convert them back to object when reading it.
 */
class PokemonTypeStringTypeConverters {

    private val gson = Gson()

    /**
     * Convert from Json [String] to [<]
     *
     * @param data json strong to convert
     * @return [<]
     */
    @TypeConverter
    fun stringToSomeObjectList(data: String?): List<Type> {
        if (data == null) {
            return emptyList()
        }

        val listType = object : TypeToken<List<Type>>() {

        }.type

        return gson.fromJson(data, listType)
    }

    /**
     * Convert from [<] to Json [String]
     *
     * @param someObjects [<] to convert
     * @return json as [String]
     */
    @TypeConverter
    fun someObjectListToString(someObjects: List<Type>): String {
        return gson.toJson(someObjects)
    }
}